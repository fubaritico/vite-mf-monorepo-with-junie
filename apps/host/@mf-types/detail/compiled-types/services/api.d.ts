export declare const fetchMovieDetail: (id?: string) => Promise<MovieDetailResponse>;
export declare const getImageUrl: (path: string, size?: string) => string;
