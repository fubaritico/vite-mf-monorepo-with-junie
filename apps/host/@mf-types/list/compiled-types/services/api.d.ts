export declare const fetchPopularMovies: (page?: number) => Promise<MovieListResponse>;
export declare const getImageUrl: (path: string, size?: string) => string;
