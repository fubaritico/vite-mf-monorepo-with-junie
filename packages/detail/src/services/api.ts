import { MovieDetailResponse } from '../types/movie';

// Note: In a real application, you would store this in an environment variable
const API_KEY = '3e837009f7b0b32d4a592ddb3c47b1c3'; // This is a dummy key, replace with a real TMDB API key
const BASE_URL = 'https://api.themoviedb.org/3';

export const fetchMovieDetail = async (id: string): Promise<MovieDetailResponse> => {
  try {
    const response = await fetch(
      `${BASE_URL}/movie/${id}?api_key=${API_KEY}&language=en-US`
    );

    if (!response.ok) {
      throw new Error(`Failed to fetch movie detail for ID: ${id}`);
    }

    return await response.json();
  } catch (error) {
    console.error(`Error fetching movie detail for ID: ${id}`, error);
    throw error;
  }
};

export const getImageUrl = (path: string, size = 'w500'): string => {
  return `https://image.tmdb.org/t/p/${size}${path}`;
};
