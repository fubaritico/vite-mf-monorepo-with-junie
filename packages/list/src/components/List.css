.movie-list {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
  gap: 2rem;
}

.movie-card {
  display: flex;
  flex-direction: column;
  background-color: #fff;
  border-radius: 8px;
  overflow: hidden;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  transition: transform 0.3s ease, box-shadow 0.3s ease;
}

.movie-card:hover {
  transform: translateY(-5px);
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
}

.movie-poster {
  position: relative;
  overflow: hidden;
  aspect-ratio: 2/3;
}

.movie-poster img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.movie-title {
  padding: 1rem;
  font-weight: 500;
  text-align: center;
  color: #333;
}

.loading, .error {
  text-align: center;
  padding: 2rem;
  font-size: 1.2rem;
}

.error {
  color: #e53935;
}
