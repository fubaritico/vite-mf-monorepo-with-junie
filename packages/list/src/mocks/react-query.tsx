import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import { useState } from 'react'

import type { RenderOptions } from '@testing-library/react'
import type { ReactNode } from 'react'

export function ReactQueryWrapper({ children }: { children: ReactNode }) {
  const [queryClient] = useState(() => new QueryClient())

  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  )
}

export const renderWithReactQuery = (
  ui: ReactNode,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: ReactQueryWrapper, ...options })
}
