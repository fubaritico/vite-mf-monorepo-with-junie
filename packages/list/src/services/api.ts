import { MovieListResponse } from '../types/movie'

const API_KEY = import.meta.env.VITE_API_KEY as string
const BASE_URL = 'https://api.themoviedb.org/3'

export const fetchPopularMovies = async (
  page = 1
): Promise<MovieListResponse> => {
  try {
    const response = await fetch(
      `${BASE_URL}/movie/popular?api_key=${API_KEY}&language=en-US&page=${page.toString()}`
    )

    if (!response.ok) {
      throw new Error('Failed to fetch popular movies')
    }

    // TODO: use react-query
    // eslint-disable-next-line
    return await response.json()
  } catch (error) {
    console.error('Error fetching popular movies:', error)
    throw error
  }
}

export const getImageUrl = (path: string, size = 'w500'): string => {
  return `https://image.tmdb.org/t/p/${size}${path}`
}
