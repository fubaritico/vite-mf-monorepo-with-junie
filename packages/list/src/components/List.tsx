import { QueryClient, useQuery } from '@tanstack/react-query'
import { Link, useLoaderData } from 'react-router-dom'

import { fetchPopularMovies, getImageUrl } from '../services/api'

import './List.css'

import type { FC } from 'react'

export type RouteComponent = FC & {
  loader: (queryClient: QueryClient) => () => Promise<MovieListResponse>
}

const query = () => ({
  queryKey: ['getMovies'],
  queryFn: async () => {
    try {
      return await fetchPopularMovies()
    } catch (error) {
      console.error('Error fetching popular movies:', error)
      throw error
    }
  },
})

const loader = (queryClient: QueryClient) => async () => {
  return queryClient.ensureQueryData(query())
}

const List: RouteComponent = () => {
  const initialData = useLoaderData<MovieListResponse>()

  const { data: movies, error } = useQuery<MovieListResponse>({
    ...query(),
    initialData,
  })

  if (error) {
    return <div className="error">{error.message}</div>
  }

  return (
    <div data-testid="movie-grid-list" className="movie-grid-list">
      {movies.results.map((movie) => (
        <Link
          to={`/detail/${movie.id.toString()}`}
          key={movie.id}
          className="movie-grid-card"
          data-testid="movie-grid-card"
        >
          <div className="movie-grid-poster">
            <img
              src={getImageUrl(movie.poster_path)}
              alt={`${movie.title} poster`}
            />
          </div>
          <div className="movie-grid-title">{movie.title}</div>
        </Link>
      ))}
    </div>
  )
}

export default List

List.loader = loader
