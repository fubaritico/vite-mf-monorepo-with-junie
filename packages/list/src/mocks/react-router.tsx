import { QueryClient } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'

import { ReactQueryWrapper } from './react-query'

import type { RouteComponent } from '../components/List'
import type { RenderOptions } from '@testing-library/react'

/**
 * Renders a component with a router provider (v7), One route config
 *
 * @param Component - the component to render and test
 * @param routes - the route to match as declared in the router configuration
 * @param path - the path to navigate to, supposed to match the route
 * @param options - additional options to pass to the render method
 * @param queryClient - the query client to use for the test
 */
export const renderWithRouter = (
  Component: RouteComponent,
  routes: string | string[] = '',
  path: string | string[] = '/',
  options?: Omit<RenderOptions, 'wrapper'>,
  queryClient = new QueryClient()
) => {
  const router = createMemoryRouter(
    Array.isArray(routes)
      ? routes.map((route) => ({
          path: route,
          element: <Component />,
          loader: Component.loader(queryClient),
        }))
      : [
          {
            path: routes,
            element: <Component />,
            loader: Component.loader(queryClient),
          },
        ],
    {
      initialEntries: Array.isArray(path) ? path : [path],
    }
  )
  return render(<RouterProvider router={router} />, {
    ...options,
  })
}

/**
 *
 * @param Component - the component to render and test
 * @param routes - the route to match as declared in the router configuration
 * @param path - the path to navigate to, supposed to match the route
 * @param options - additional options to pass to the render method
 */
export const renderReactQueryWithRouter = (
  Component: RouteComponent,
  routes: string | string[] = '',
  path: string | string[] = '/',
  options?: RenderOptions
) => {
  return renderWithRouter(Component, routes, path, {
    wrapper: ReactQueryWrapper,
    ...options,
  })
}
