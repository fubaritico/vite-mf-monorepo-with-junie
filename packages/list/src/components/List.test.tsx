import { cleanup, screen, waitFor, within } from '@testing-library/react'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

import moviesData from '../mocks/data/popular'
import { renderReactQueryWithRouter } from '../mocks/react-router'
import * as mod from '../services/api'

import List from './List'

// Spying to assert calls to the API
const fetchPopularMovies = vi.spyOn(mod, 'fetchPopularMovies')
const getImageUrl = vi.spyOn(mod, 'getImageUrl')

// fetchPopularMovies locale mock
vi.mock('../services/api', async (importOriginal) => ({
  ...(await importOriginal<typeof import('../services/api')>()),
  fetchPopularMovies: vi.fn(async () => Promise.resolve(moviesData)),
}))

/**
 * Helper function to flush promises, so that any pending promises are resolved
 * and the component is properly rendered before the next assertion.
 */
function flushPromises() {
  return new Promise((resolve) => {
    setTimeout(resolve, 0)
  })
}

describe('List', () => {
  beforeEach(() => {
    fetchPopularMovies.mockClear()
    getImageUrl.mockClear()
  })

  // cleanup after each test, vitest doesn't perform any automatic DOM cleanup
  afterEach(() => {
    cleanup()
  })

  /**
   * Because the init is async, some extra assertion is to be performed to ensure
   * there are no memory leaks between tests regarding spy clearing.
   */
  it('should render the list of movies', async () => {
    const { container } = renderReactQueryWithRouter(List)
    // testing API call on init
    expect(fetchPopularMovies).toHaveBeenCalledTimes(1)
    // testing any rendered element to be sure that the component is mounted and rendered
    const element = await waitFor(() => screen.getByText('Havoc'))

    expect(element).toBeTruthy()
    expect(container).toMatchSnapshot()

    await flushPromises()
  })

  it('should call "fetchPopularMovies" one time on init', async () => {
    renderReactQueryWithRouter(List)

    expect(fetchPopularMovies).toHaveBeenCalledTimes(1)

    await flushPromises()
  })

  // Testing the same to verify that each test is properly isolated
  it('should call "fetchPopularMovies" one time on init a second time', async () => {
    renderReactQueryWithRouter(List)

    expect(fetchPopularMovies).toHaveBeenCalledTimes(1)

    await flushPromises()
  })

  it('should display the right amount of cards in the movie grid', async () => {
    renderReactQueryWithRouter(List)

    expect(fetchPopularMovies).toHaveBeenCalledTimes(1)

    await waitFor(() => {
      expect(screen.getAllByTestId('movie-grid-card').length).toEqual(
        moviesData.results.length
      )
    })

    await flushPromises()
  })

  it.each([
    ...moviesData.results.map(({ title, poster_path }, i) => [
      title,
      'https://image.tmdb.org/t/p/w500' + poster_path,
      i,
    ]),
  ])(
    'should display a card with a title "%s" and a poster with src equals to "%s"',
    async (title, poster_path, index) => {
      renderReactQueryWithRouter(List)

      expect(fetchPopularMovies).toHaveBeenCalledTimes(1)

      await waitFor(() => {
        expect(
          screen.getAllByTestId('movie-grid-card')[index as number]
        ).toBeTruthy()
      })

      expect(getImageUrl).toHaveBeenCalledTimes(moviesData.results.length)
      expect(getImageUrl).toHaveBeenCalledWith(
        moviesData.results[index as number].poster_path
      )

      const card = screen.getAllByTestId('movie-grid-card')[index as number]

      expect(within(card).getByText(title)).toBeTruthy()
      expect(within(card).getByRole('img').getAttribute('src')).toBe(
        poster_path
      )

      await flushPromises()
    }
  )
})
