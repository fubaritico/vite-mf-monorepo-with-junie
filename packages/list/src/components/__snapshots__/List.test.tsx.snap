// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List > should render the list of movies 1`] = `
<div>
  <div
    class="movie-grid-list"
    data-testid="movie-grid-list"
  >
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/668489"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Havoc poster"
          src="https://image.tmdb.org/t/p/w500/r46leE6PSzLR3pnVzaxx5Q30yUF.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Havoc
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1197306"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="A Working Man poster"
          src="https://image.tmdb.org/t/p/w500/xUkUZ8eOnrOnnJAfusZUqKYZiDu.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        A Working Man
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1233069"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Exterritorial poster"
          src="https://image.tmdb.org/t/p/w500/qWTfHG8KdXwr0bqypYbNZLenh0J.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Exterritorial
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/950387"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="A Minecraft Movie poster"
          src="https://image.tmdb.org/t/p/w500/iPPTGh2OXuIv6d7cwuoPkw8govp.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        A Minecraft Movie
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1471014"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Van Gogh by Vincent poster"
          src="https://image.tmdb.org/t/p/w500/z73X4WKZghBh5fri31o8P6vBEB2.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Van Gogh by Vincent
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1276073"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Bullet Train Explosion poster"
          src="https://image.tmdb.org/t/p/w500/qkTKtOHK9JEEOHgPQZ0dFtzs5ML.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Bullet Train Explosion
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/324544"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="In the Lost Lands poster"
          src="https://image.tmdb.org/t/p/w500/dDlfjR7gllmr8HTeN6rfrYhTdwX.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        In the Lost Lands
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1225915"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Jewel Thief: The Heist Begins poster"
          src="https://image.tmdb.org/t/p/w500/eujLbO0kf1eqWC8XpHUJdtAVW2J.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Jewel Thief: The Heist Begins
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/986056"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Thunderbolts* poster"
          src="https://image.tmdb.org/t/p/w500/hqcexYHbiTBfDIdDWxrxPtVndBX.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Thunderbolts*
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/822119"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Captain America: Brave New World poster"
          src="https://image.tmdb.org/t/p/w500/pzIddUEMWhWzfvLI3TwxUG2wGoi.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Captain America: Brave New World
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1180906"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Desert Dawn poster"
          src="https://image.tmdb.org/t/p/w500/S21BfLrJSD9njucBfY3CKqp8rD.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Desert Dawn
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1233413"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Sinners poster"
          src="https://image.tmdb.org/t/p/w500/jYfMTSiFFK7ffbY2lay4zyvTkEk.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Sinners
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1124620"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="The Monkey poster"
          src="https://image.tmdb.org/t/p/w500/yYa8Onk9ow7ukcnfp2QWVvjWYel.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        The Monkey
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1153714"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Death of a Unicorn poster"
          src="https://image.tmdb.org/t/p/w500/lXR32JepFwD1UHkplWqtBP1K79z.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Death of a Unicorn
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1092899"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="The Siege poster"
          src="https://image.tmdb.org/t/p/w500/hVh4hMzkXNLnScudbid6hDvjMPk.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        The Siege
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/1414048"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Day of Reckoning poster"
          src="https://image.tmdb.org/t/p/w500/cVSjSQryFUERYSdOmdgJ2m0eBte.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Day of Reckoning
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/348"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Alien poster"
          src="https://image.tmdb.org/t/p/w500/vfrQk5IPloGg1v9Rzbh2Eg3VGyM.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Alien
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/335984"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Blade Runner 2049 poster"
          src="https://image.tmdb.org/t/p/w500/gajva2L0rPYkEWjzgFlBXCAVBE5.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Blade Runner 2049
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/931349"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Ash poster"
          src="https://image.tmdb.org/t/p/w500/5Oz39iyRuztiA8XqCNVDBuy2Ut3.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Ash
      </div>
    </a>
    <a
      class="movie-grid-card"
      data-discover="true"
      data-testid="movie-grid-card"
      href="/detail/374720"
    >
      <div
        class="movie-grid-poster"
      >
        <img
          alt="Dunkirk poster"
          src="https://image.tmdb.org/t/p/w500/b4Oe15CGLL61Ped0RAS9JpqdmCt.jpg"
        />
      </div>
      <div
        class="movie-grid-title"
      >
        Dunkirk
      </div>
    </a>
  </div>
</div>
`;
